Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)

I am attempting to build the AMReX library on development branch (16714342b3) cloned this morning with default build options. I am doing this by `./configure && make`. My version of MPI is MPICH 3.3.2 installed from the package manager of Fedora 33. The build fails in AMReX Fortran codes, output follows. Is the makefile build still supported?

```
mpif90 -g -O3 -ffree-line-length-none -fno-range-check -fno-second-underscore -fimplicit-none   -Jtmp_build_dir/o/3d.gnu.MPI.EXE -I tmp_build_dir/o/3d.gnu.MPI.EXE -DBL_LANG_FORT -DAMREX_LANG_FORT -DBL_USE_MPI -DAMREX_USE_MPI -DAMREX_GIT_VERSION=\"21.05-43-g16714342b3c0\" -DBL_SPACEDIM=3 -DAMREX_SPACEDIM=3 -DBL_FORT_USE_UNDERSCORE -DAMREX_FORT_USE_UNDERSCORE -DBL_Linux -DAMREX_Linux -DAMREX_PARTICLES -DNDEBUG -DOMPI_SKIP_MPICXX -DAMREX_PARTICLES -DBL_USE_F_INTERFACES -Itmp_build_dir/s/3d.gnu.MPI.EXE -I. -I/work/SENSEI/amrex/Src/Base -I/work/SENSEI/amrex/Src/AmrCore -I/work/SENSEI/amrex/Src/Amr -I/work/SENSEI/amrex/Src/Boundary -I/work/SENSEI/amrex/Src/Particle -I/work/SENSEI/amrex/Src/F_Interfaces/Particle -I/work/SENSEI/amrex/Src/F_Interfaces/Base -I/work/SENSEI/amrex/Src/F_Interfaces/Octree -I/work/SENSEI/amrex/Src/F_Interfaces/AmrCore -I/work/SENSEI/amrex/Src/LinearSolvers/MLMG -I/work/SENSEI/amrex/Src/LinearSolvers/Projections -I/work/SENSEI/amrex/Src/F_Interfaces/LinearSolvers  -c /work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90 -o tmp_build_dir/o/3d.gnu.MPI.EXE/AMReX_parallel_mod.o
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:294:23:

  294 |        call MPI_Reduce(tmp, r, 1, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:296:26:

  296 |        call MPI_Allreduce(tmp, r, 1, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:279:23:

  279 |        call MPI_Reduce(tmp, i, n, MPI_INTEGER, MPI_MIN, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:281:26:

  281 |        call MPI_Allreduce(tmp, i, n, MPI_INTEGER, MPI_MIN, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:264:23:

  264 |        call MPI_Reduce(tmp, i, 1, MPI_INTEGER, MPI_MIN, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:266:26:

  266 |        call MPI_Allreduce(tmp, i, 1, MPI_INTEGER, MPI_MIN, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:234:23:

  234 |        call MPI_Reduce(tmp, r, 1, amrex_mpi_real, MPI_MAX, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:236:26:

  236 |        call MPI_Allreduce(tmp, r, 1, amrex_mpi_real, MPI_MAX, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:219:23:

  219 |        call MPI_Reduce(tmp, i, n, MPI_INTEGER, MPI_MAX, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:221:26:

  221 |        call MPI_Allreduce(tmp, i, n, MPI_INTEGER, MPI_MAX, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:204:23:

  204 |        call MPI_Reduce(tmp, i, 1, MPI_INTEGER, MPI_MAX, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:206:26:

  206 |        call MPI_Allreduce(tmp, i, 1, MPI_INTEGER, MPI_MAX, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:174:23:

  174 |        call MPI_Reduce(tmp, r, 1, amrex_mpi_real, MPI_SUM, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:176:26:

  176 |        call MPI_Allreduce(tmp, r, 1, amrex_mpi_real, MPI_SUM, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:159:23:

  159 |        call MPI_Reduce(tmp, i, n, MPI_INTEGER, MPI_SUM, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:161:26:

  161 |        call MPI_Allreduce(tmp, i, n, MPI_INTEGER, MPI_SUM, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:144:23:

  144 |        call MPI_Reduce(tmp, i, 1, MPI_INTEGER, MPI_SUM, rank, m_comm, ierr)
      |                       1
......
  310 |        call MPI_Reduce(tmp, r, n, amrex_mpi_real, MPI_MIN, rank, m_comm, ierr)
      |                       2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
/work/SENSEI/amrex/Src/F_Interfaces/Base/AMReX_parallel_mod.F90:146:26:

  146 |        call MPI_Allreduce(tmp, i, 1, MPI_INTEGER, MPI_SUM, m_comm, ierr)
      |                          1
......
  312 |        call MPI_Allreduce(tmp, r, n, amrex_mpi_real, MPI_MIN, m_comm, ierr)
      |                          2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
make: *** [/work/SENSEI/amrex/Tools/GNUMake/Make.rules:262: tmp_build_dir/o/3d.gnu.MPI.EXE/AMReX_parallel_mod.o] Error 1
```

--- a/Tools/GNUMake/comps/gnu.mak
+++ b/Tools/GNUMake/comps/gnu.mak
@@ -212,6 +212,15 @@ F90FLAGS += -ffree-line-length-none -fno-range-check -fno-second-underscore -fim
 
 FMODULES =  -J$(fmoddir) -I $(fmoddir)
 
+# gcc 10 has treated mismatches between actuall and dummy argument lists
+# in a single file as errors. This is a workaround.
+ifeq ($(USE_MPI),TRUE)
+ifeq ($(gcc_major_ge_10),1)
+  F90FLAGS += -fallow-argument-mismatch
+  FFLAGS   += -fallow-argument-mismatch
+endif
+endif
+
 ########################################################################
 
 ifneq ($(BL_NO_FORT),TRUE)

--- a/Tools/GNUMake/comps/hip.mak
+++ b/Tools/GNUMake/comps/hip.mak
@@ -29,6 +29,15 @@ F90FLAGS := -ffree-line-length-none -fno-range-check -fno-second-underscore -fim
 
 FMODULES =  -J$(fmoddir) -I $(fmoddir)
 
+ifeq ($(USE_MPI),TRUE)
+gfortran_major_version = $(shell gfortran -dumpfullversion -dumpversion | head -1 | sed -e 's;.*  *;;' | sed -e 's;\..*;;')
+gfortran_major_ge_10 = $(shell expr $(gfortran_major_version) \>= 10)
+ifeq ($(gfortran_major_ge_10),1)
+  F90FLAGS += -fallow-argument-mismatch
+  FFLAGS   += -fallow-argument-mismatch
+endif
+endif
+
 # rdc support
 CXXFLAGS += -fgpu-rdc
 HIPCC_FLAGS += -fgpu-rdc # This will be added to link flags

--- a/Tools/GNUMake/comps/llvm.mak
+++ b/Tools/GNUMake/comps/llvm.mak
@@ -78,6 +78,15 @@ F90FLAGS += -ffree-line-length-none -fno-range-check -fno-second-underscore -fim
 
 FMODULES =  -J$(fmoddir) -I $(fmoddir)
 
+ifeq ($(USE_MPI),TRUE)
+gfortran_major_version = $(shell gfortran -dumpfullversion -dumpversion | head -1 | sed -e 's;.*  *;;' | sed -e 's;\..*;;')
+gfortran_major_ge_10 = $(shell expr $(gfortran_major_version) \>= 10)
+ifeq ($(gfortran_major_ge_10),1)
+  F90FLAGS += -fallow-argument-mismatch
+  FFLAGS   += -fallow-argument-mismatch
+endif
+endif
+
 ########################################################################
 
 GENERIC_COMP_FLAGS =

